src/mobu/jupyterclient.py

```diff
@@ -158,98 +165,130 @@ async def lab_login(self) -> None:
+        # the spawn-pending page, which will return a 200.
+        data = self.jupyter_options_form
+        self.log.info(f"Spawning lab for {self.user.username}")
+        async with self.session.post(spawn_url, data=data) as r:
+            if r.status != 200:
+                msg = "Unexpected reply status after spawning"
+                await self._raise_error(msg, r)
 
-        # Jupyterlab will give up a spawn after 900 seconds, so we shouldn't
-        # wait longer than that.
+        # Poll until the lab has spawned.  Jupyterlab will give up a spawn
+        # after 900 seconds, so we shouldn't wait longer than that.
         max_poll_secs = 900
         poll_interval = 15
         retries = max_poll_secs / poll_interval
-
+        start = datetime.now(tz=timezone.utc)
         while retries > 0:
-            async with self.session.get(hub_url) as r:
-                if str(r.url) == lab_url:
-                    self.log.info(f"Lab spawned, redirected to {r.url}")
-                    return
-
-                if not r.ok:
-                    await self._raise_error("Error spawning", r)
+            if await self.lab_has_spawned():
+                return
+            now = datetime.now(tz=timezone.utc)
+            elapsed = int((now - start).total_seconds())
+            self.log.info(f"Waiting for lab to spawn ({elapsed}s elapsed)")
+            retries -= 1
+            await asyncio.sleep(poll_interval)
 
-                self.log.info(f"Still waiting for lab to spawn [{r.status}]")
-                retries -= 1
-                await asyncio.sleep(poll_interval)
-
-        raise Exception("Giving up waiting for lab to spawn!")
+        # Timed out spawning the lab.
+        raise LabSpawnTimeoutError("Lab did not spawn after {max_poll_secs}s")
 
     async def delete_lab(self) -> None:
-        if not await self.is_lab_running():
+        if await self.lab_is_stopped():
+            self.log.info("Lab is already stopped")
             return
 
-        server_url = (
-            self.jupyter_url + f"hub/api/users/{self.user.username}/server"
-        )
-        self.log.info(f"Deleting lab for {self.user.username} at {server_url}")
+        user = self.user.username
+        server_url = self.jupyter_url + f"hub/api/users/{user}/server"
         headers = {"Referer": self.jupyter_url + "hub/home"}
         async with self.session.delete(server_url, headers=headers) as r:
             if r.status not in [200, 202, 204]:
                 await self._raise_error("Error deleting lab", r)
 
         # Wait for the lab to actually go away.  If we don't do this, we may
         # try to create a new lab while the old one is still shutting down.
-        count = 0
-        while await self.is_lab_running() and count < 10:
-            self.log.info(f"Waiting for lab deletion ({count}s elapsed)")
-            await asyncio.sleep(1)
-            count += 1
-        if await self.is_lab_running():
+        max_poll_secs = 20
```

Comment:
I would probably make this 30, just because I'm a little twitchy about what's going to happen on underpowered T&S teststands.

Reply:
I want to figure out how to make more of these waits interruptable and then it will be less worrisome to increase them. FWIW, on the IDF the deletion usually happens within 5s, but agreed that underpowered teststands could be a problem.